fix: disallow AlterColumn method chaining

soanvig opened the pull request:
Fixes: #467

It was slightly tricky, but I decided that "unwinding" looks the best in terms of clarity of what is happening.

This effectively produces query:

```
ALTER TABLE "x" ALTER COLUMN "y" SET DEFAULT true, ALTER COLUMN "z" DROP NOT NULL, [...]
```

as in Postgres docs. MySQL has exact syntax for chaining ALTER COLUMN, but supports only `drop default` and `set default` operations, so probably nobody will chain those two together, plus alter column usage for MySQL is not properly handled in whole Kysely anyway. I tested MySQL case manually.

---

EDIT: now it just disallows chaining

igalklebanov commented:
Hey 👋 

I understand what you were going for with this and it does seem convenient, but it is not aligned with Kysely's WYSIWYG design principle. Not sure about these changes.

We should consider either:

1. each method returns a builder that cannot chain further and `.alterColumn` expects it as argument. So the original issue is blocked by more precise API.

2. make compiler just add a whitespace between each operation, and keep it a single `alter column` clause. This'll cause a runtime error obviously with PostgreSQL/MySQL - but consumer should know SQL. This might provide future-proofing in case some dialect supports this.

igalklebanov commented:
> plus alter column usage for MySQL is not properly handled in whole Kysely anyway. I tested MySQL case manually.

Can you explain this? does it require an issue?

soanvig commented:
@igalklebanov  that's rather new to me that Kysely pushes WYSIWYG to such extent.
I can wait for other people decision on that topic, sure.

---

> Can you explain this? does it require an issue?

No. I was just saying, that you can call `AlterTableBuilder.dropNotNull` even if it is not supported in MySQL (and you will get runtime error)

igalklebanov commented:
> No. I was just saying, that you can call AlterTableBuilder.dropNotNull even if it is not supported in MySQL (and you will get runtime error)

Oh, gotcha. The API is dialect agnostic by design.

koskimas commented:
> each method returns a builder that cannot chain further and .alterColumn expects it as argument. So the original issue is blocked by more precise API.

@igalklebanov Yes, I agree. This is the way to proceed.

soanvig commented:
@koskimas @igalklebanov 
So for fun I'll implement that part, and we'll see. I really like this library, and we heavily use it at work, so I'm glad to contribute something.

soanvig commented:
@igalklebanov 

I updated the PR. Above you can see intelisense suggestion (type compilation result).
Besides restricting the result, I also narrowed API of `AlterColumnNode`, so it cannot contain more than one operation at once (since it won't be handled properly anyway): I removed `cloneWith`, because it was not needed, and creating the node now requires passing the property to be set.

igalklebanov commented on src/operation-node/alter-column-node.ts:
```diff
@@ -26,20 +24,11 @@ export const AlterColumnNode = freeze({
   },
 
-  create(column: string): AlterColumnNode {
+  create<T extends keyof AlterColumnNode>(column: string, prop: T, value: Required<AlterColumnNode>[T]): AlterColumnNode {
```
You're allowing setting `'kind'` and `'column'` as prop.

```suggestion
  create<P extends keyof AlterColumnNodeProps>(column: string, prop: P, value: Required<AlterColumnNode>[P]): AlterColumnNode {
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -10,49 +10,39 @@ import {
 
 export class AlterColumnBuilder {
-  protected readonly alterColumnNode: AlterColumnNode
+  protected column: string
```
```suggestion
  readonly #column: string
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -10,49 +10,39 @@ import {
-  constructor(alterColumnNode: AlterColumnNode) {
-    this.alterColumnNode = alterColumnNode
+  constructor(column: string) {
+    this.column = column
```
```suggestion
    this.#column = column
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -10,49 +10,39 @@ import {
-      AlterColumnNode.cloneWith(this.alterColumnNode, {
-        dataType: parseDataTypeExpression(dataType),
-      })
+      AlterColumnNode.create(this.column, 'dataType', parseDataTypeExpression(dataType))
```
```suggestion
      AlterColumnNode.create(this.#column, 'dataType', parseDataTypeExpression(dataType))
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -10,49 +10,39 @@ import {
-      AlterColumnNode.cloneWith(this.alterColumnNode, {
-        setDefault: parseDefaultValueExpression(value),
-      })
+      AlterColumnNode.create(this.column, 'setDefault', parseDefaultValueExpression(value))
```
```suggestion
      AlterColumnNode.create(this.#column, 'setDefault', parseDefaultValueExpression(value))
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -10,49 +10,39 @@ import {
-      AlterColumnNode.cloneWith(this.alterColumnNode, {
-        dropDefault: true,
-      })
+      AlterColumnNode.create(this.column, 'dropDefault', true)
```
```suggestion
      AlterColumnNode.create(this.#column, 'dropDefault', true)
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -10,49 +10,39 @@ import {
-      AlterColumnNode.cloneWith(this.alterColumnNode, {
-        setNotNull: true,
-      })
+      AlterColumnNode.create(this.column, 'setNotNull', true)
```
```suggestion
      AlterColumnNode.create(this.#column, 'setNotNull', true)
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -10,49 +10,39 @@ import {
-      AlterColumnNode.cloneWith(this.alterColumnNode, {
-        dropNotNull: true,
-      })
+      AlterColumnNode.create(this.column, 'dropNotNull', true)
```
```suggestion
      AlterColumnNode.create(this.#column, 'dropNotNull', true)
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -66,20 +56,30 @@ export class AlterColumnBuilder {
 
 /**
- * Allows us to force consumers to do something, anything, when altering a column.
+ * Allows us to force consumers to do something, anything, when altering a column,
+ * and also disallows chaining more methods on AlterColumnBuilder, because
+ * SQL syntax for chaining ALTER COLUMN operations would be not WYSIWYG
```
I think the addition is TMI.

```suggestion
 * Allows us to force consumers to do exactly one alteration to a column.
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -66,20 +56,30 @@ export class AlterColumnBuilder {
-  implements OperationNodeSource
+export class AlteredColumnBuilder implements OperationNodeSource
 {
+  protected readonly alterColumnNode: AlterColumnNode
```
```suggestion
  readonly #alterColumnNode: AlterColumnNode
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -66,20 +56,30 @@ export class AlterColumnBuilder {
+  protected readonly alterColumnNode: AlterColumnNode
+
+  constructor(alterColumnNode: AlterColumnNode) {
+    this.alterColumnNode = alterColumnNode
```
```suggestion
    this.#alterColumnNode = alterColumnNode
```

igalklebanov commented on src/schema/alter-column-builder.ts:
```diff
@@ -66,20 +56,30 @@ export class AlterColumnBuilder {
+  }
+
   toOperationNode(): AlterColumnNode {
     return this.alterColumnNode
```
```suggestion
    return this.#alterColumnNode
```

igalklebanov reviewed (commented):
Looks good! 💪 

Left a few minor suggestions..

soanvig commented:
@igalklebanov thanks for the suggestions. I've implemented them all in one commit.

igalklebanov reviewed (approved):
LGTM! 🚀
